### PR TITLE
fix(eth-flow): additionally check eth-flow order id in smart-contract

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Issue:
https://docs.google.com/document/d/1sIwsE9sCuEMZoNuBVXxpdxLrW-s-tF75AEnAMUX05I0/edit#heading=h.z0cmq0qselmp

I've checked the system based on `ethFlowInFlightOrderIdsAtom` and it works well.
Anyway, I've added an additional check with static call `orders[orderDigest]`.

  # To Test

I wasn't able to reproduce the issue because the previous solution worked well. But, I've mocked it and checked that the new approach with static call also works.

### CODE REVIEW

File: libs/abis/src/abis/CoWSwapEthFlow.json [hunk 1]
```diff
@@ -0,0 +1,152 @@
+[
```
Comment (shoom3301) [inline libs/abis/src/abis/CoWSwapEthFlow.json:1]: Trimmed and moved `CoWSwapEthFlow.json` to our repo.

Old file size: 389kb
New file size: 3.4kb

File: apps/cowswap-frontend/src/modules/swap/hooks/useEthFlowContext.ts [hunk 2]
```diff
@@ -9,23 +8,18 @@ import { useTransactionAdder } from 'legacy/state/enhancedTransactions/hooks'
...
   const addTransaction = useTransactionAdder()
 
-  const ethFlowInFlightOrderIds = useAtomValue(ethFlowInFlightOrderIdsAtom)
   const addInFlightOrderId = useSetAtom(addInFlightOrderIdAtom)
 
-  const checkInFlightOrderIdExists = useCallback(
-    (orderId: string) => ethFlowInFlightOrderIds.includes(orderId),
-    [ethFlowInFlightOrderIds]
-  )
+  const isEthFlowOrderExists = useIsEthFlowOrderExists()
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/hooks/useEthFlowContext.ts:22]: The variable name indicates to me this is a boolean.
Reading the code though, it's a callback.

I suggest renaming the hook and callback to:

useCheckEthFlowOrderExists
checkEthFlowOrderExists

File: apps/cowswap-frontend/src/modules/swap/hooks/useIsEthFlowOrderExists.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    async (orderId: string, orderDigest: string) => {
+      if (ethFlowInFlightOrderIds.includes(orderId)) {
+        return true
+      }
+
+      if (ethFlowContract) {
+        try {
+          const { owner } = await ethFlowContract.callStatic.orders(orderDigest)
+
+          return owner.toLowerCase() !== ZERO_ADDRESS
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/hooks/useIsEthFlowOrderExists.ts:27]: Don't we need to check the owner matches the one placing the order?

File: apps/cowswap-frontend/src/legacy/hooks/useContract.ts [hunk 4]
```diff
@@ -9,7 +9,6 @@ import {
   Erc721,
   Erc1155,
   Weth,
-  CoWSwapEthFlowJson,
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/legacy/hooks/useContract.ts:12]: Shouldn't it be removed from the abis app? https://github.com/cowprotocol/cowswap/blob/8365242936104e23bba14d970ff6256c0ea55544/libs/abis/src/index.ts#L23

Comment (shoom3301) [reply apps/cowswap-frontend/src/legacy/hooks/useContract.ts:12]: Thanks! Deleted

File: apps/cowswap-frontend/src/modules/swap/hooks/useIsEthFlowOrderExists.ts (hunk 3 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/hooks/useIsEthFlowOrderExists.ts:27]: Yes and no at the same time :)
If the smart-contract storage contains an order with the same id, it means there is a collision.
How orders map looks like:
```
mapping(bytes32 => EthFlowOrder.OnchainData) public orders;
```

It's actually key-value, where key is `orderDigest`. And here I see smth interesting, `orderDigest` is a hash of an order that doesn't contain the owner address implicitly. It contains it as a `receiver` but the receiver can be any address.
So, theoretically, two different accounts can create an order with the same `orderDigest` and in this key the last one will face a collision.
In conclusion, I think we should take into account even this case and check if `owner` is not zero.

image

Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/hooks/useIsEthFlowOrderExists.ts:27]:
```
if (orders[orderHash].owner != EthFlowOrder.NO_OWNER) {
            revert OrderIsAlreadyOwned(orderHash);
        }
```

Even in the smart-contact they do the same check

File: apps/cowswap-frontend/src/modules/swap/hooks/useEthFlowContext.ts (hunk 2 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/hooks/useEthFlowContext.ts]: Thanks! Renamed

**Review by alfetopito — APPROVED** [review]
I tested and it worked fine.
Placed 2 eth flow orders in sequence before the first was mined.
Both were created with distinct order ids

https://dev.explorer.cow.fi/goerli/orders/0x7685944627474cb4d66b2fc9eca520a0540fc114bcf5db7c2fa120bc8e76d54ad02de8da0b71e1b59489794f423fabba2adc4d93ffffffff?tab=overview
https://dev.explorer.cow.fi/goerli/orders/0x63d3d2fa17fbc58c5c394c23bdcbb52871aa855291f8a6280182cb3648916481d02de8da0b71e1b59489794f423fabba2adc4d93ffffffff?tab=overview

Also, can you point this to the release branch?

File: apps/cowswap-frontend/src/modules/swap/hooks/useCheckEthFlowOrderExists.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+export function useCheckEthFlowOrderExists(): EthFlowOrderExistsCallback {
+  const ethFlowInFlightOrderIds = useAtomValue(ethFlowInFlightOrderIdsAtom)
+  const ethFlowContract = useEthFlowContract()
+
+  return useCallback(
+    async (orderId: string, orderDigest: string) => {
+      if (ethFlowInFlightOrderIds.includes(orderId)) {
+        return true
+      }
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/swap/hooks/useCheckEthFlowOrderExists.ts:21]: cool, it still uses the inFlight ids :) 

this will make it very unlikely, but at the same time we will do it when we are unsure. Thats great!

File: apps/cowswap-frontend/src/modules/swap/hooks/useCheckEthFlowOrderExists.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+      if (ethFlowContract) {
+        try {
+          const { owner } = await ethFlowContract.callStatic.orders(orderDigest)
+
+          return owner.toLowerCase() !== ZERO_ADDRESS
+        } catch (e) {
+          console.error('Eth-flow order existing check error: ', e)
+
+          return false
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/swap/hooks/useCheckEthFlowOrderExists.ts:31]: So if we have an error, we take a bet :) 

okok, i guess is better than not continuing with the flow, and re-attempts might be an overkill.

File: libs/abis/src/abis/CoWSwapEthFlow.json (hunk 1 above)
Comment (anxolin) [reply libs/abis/src/abis/CoWSwapEthFlow.json:1]: That's a big difference!

**Review by anxolin — APPROVED** [review]
Great PR
